### PR TITLE
Images processing

### PR DESCRIPTION
**Pull request by blocknotes**
This PR creates a product/variant processor that saves the entity images.

### CODE REVIEW

File: lib/solidus_importer/processors/product_images.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        @data = context[:data]
+        @entity = context[:entity]
+        context.merge!(check_data || save_images)
+      end
+
+      private
+
+      def check_data
+        if @data.blank?
+          { success: false, messages: 'Missing input data' }
```
Comment (elia) [inline lib/solidus_importer/processors/product_images.rb]: Can't we assume we have _some_ data here and remove this check?

File: lib/solidus_importer/processors/product_images.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+module SolidusImporter
+  module Processors
+    class ProductImages < Base
+      def call(context)
+        @data = context[:data]
+        @entity = context[:entity]
```
Comment (elia) [inline lib/solidus_importer/processors/product_images.rb]: Since this will necessarily be working on a product wouldn't it be more clear if the context key was `:product` instead of `:entity`? Also would suggest using `.fetch` where we require something to be there.

File: lib/solidus_importer/processors/product_images.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+      private
+
+      def check_data
+        if @data.blank?
+          { success: false, messages: 'Missing input data' }
+        elsif @data['Image Src'].blank?
+          {}
+        elsif !@entity || !@entity.is_a?(Spree::Product) || !@entity.valid?
+          { success: false, messages: 'Target entity must be a valid product' }
```
Comment (elia) [inline lib/solidus_importer/processors/product_images.rb]: As above, checking for the product's validity here seems out of scope.

File: lib/solidus_importer/processors/product_images.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        end
+      end
+
+      def prepare_image
+        Spree::Image.new(
+          attachment: File.open(@data['Image Src']),
```
Comment (elia) [inline lib/solidus_importer/processors/product_images.rb]: Are we positive that the `Image Src` is a file path? I would have expected it to be a url instead…

File: lib/solidus_importer/processors/product_images.rb (hunk 1 above)
Comment (blocknotes) [reply lib/solidus_importer/processors/product_images.rb]: Good point 😄
I'm removing it (and the same for variant related one).

File: lib/solidus_importer/processors/product_images.rb (hunk 2 above)
Comment (blocknotes) [reply lib/solidus_importer/processors/product_images.rb]: > Since this will necessarily be working on a product wouldn't it be more clear if the context key was `:product` instead of `:entity`?

I'm not completely convinced about this one: `:entity` is already prepared by the Base processor so for me it's like a standard. Like: _every processor that imports something will create the entity key_.

Otherwise, I should create a key for each type, in this case, the Product processor should create `:product` in place of `:entity`? The pro here is that I could avoid the check about the entity as Product...

> Also would suggest using `.fetch` where we require something to be there.

The problem that I see about the `.fetch` method is that I should pass a context object instead of a plain hash. Another bad point for me is that if a context key is also modified I should supply an `#apply` method to save in the context (i.e. the `:data` key)

File: lib/solidus_importer/processors/product_images.rb (hunk 4 above)
Comment (blocknotes) [reply lib/solidus_importer/processors/product_images.rb]: You are right 😅
What do you think it's better? To assume that we will always have remote URLs? To look for HTTP urls otherwise the resource is local?

File: lib/solidus_importer/processors/product_images.rb (hunk 1 above)
Comment (blocknotes) [reply lib/solidus_importer/processors/product_images.rb]: Done, commits:
https://github.com/nebulab/solidus_importer/pull/8/commits/b2a76cc6776f5fee41fead98c412888bc363dd03
https://github.com/nebulab/solidus_importer/pull/8/commits/7a0cdd322f602d21cd697174c4ca20db39af4511

File: lib/solidus_importer/processors/product_images.rb (hunk 4 above)
Comment (elia) [reply lib/solidus_importer/processors/product_images.rb]: Probably safe to always expect URLs and leave local files for customizations 👍

File: lib/solidus_importer/processors/product_images.rb (hunk 2 above)
Comment (blocknotes) [reply lib/solidus_importer/processors/product_images.rb]: After a call with @elia, I agree with him and will make some changes as proposed by him

File: lib/solidus_importer/processors/variant.rb [hunk 5]
```diff
@@ -14,10 +14,10 @@ def call(context)
       def check_data
         if @data.blank?
           { success: false, messages: 'Missing input data' }
-        elsif @data['Variant SKU'].blank?
+        elsif !variant?
           {}
-        elsif !@product
-          { success: false, messages: 'Missing product' }
+        elsif !@product || !@product.valid?
```
Comment (elia) [inline lib/solidus_importer/processors/variant.rb]: Might be more readable as `!@product&.valid?`

File: lib/solidus_importer/processors/variant.rb [hunk 6]
```diff
@@ -33,6 +33,10 @@ def save_variant
         variant = prepare_variant
         prepare_context(entity: variant, new_record: variant.new_record?, success: variant.save)
       end
+
+      def variant?
+        @variant ||= (@data && @data['Variant SKU'].present?)
```
Comment (elia) [inline lib/solidus_importer/processors/variant.rb]: Why checking for the presence of `@data` here and in other place? 
I thought we expected it to be there, is there a scenario in which it can end up being nil or blank?

Comment (blocknotes) [reply lib/solidus_importer/processors/variant.rb]: Because `@data.blank?` check is made after the line `@product = context.fetch(:product) if variant?` - perhaps it looks a bit weird.

Comment (blocknotes) [reply lib/solidus_importer/processors/variant.rb]: Updated to remove data empty check from processors

File: lib/solidus_importer/processors/product_images.rb (hunk 3 above)
Comment (elia) [reply lib/solidus_importer/processors/product_images.rb]: @blocknotes remember to fix this one 🎗️

Comment (blocknotes) [reply lib/solidus_importer/processors/product_images.rb]: I fixed it here:
https://github.com/nebulab/solidus_importer/pull/8/commits/8173598e686c19a2f1b4b59e0e201a48cc7e9595

Comment (blocknotes) [reply lib/solidus_importer/processors/product_images.rb]: Sure 😅 done now. Here:
https://github.com/nebulab/solidus_importer/pull/8/commits/85b45961869aa7f1b65acf1dfc2d64eba11958b3
